internal/controller/resources/resources.go

```diff
@@ -0,0 +1,113 @@
+package resources
+
+import (
+	"context"
+	goerrors "errors"
+	"fmt"
+
+	"github.com/dana-team/nfspvc-operator/internal/controller/finalizer"
+	"github.com/dana-team/nfspvc-operator/internal/controller/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	danaiov1alpha1 "github.com/dana-team/nfspvc-operator/api/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var FailedCleanupError = goerrors.New("failed nfspvc cleanup")
+
+// HandleDelete ensures the deletion of the nfspvc.
+func HandleDelete(ctx context.Context, nfspvc danaiov1alpha1.NfsPvc, k8sClient client.Client) (error, bool) {
+	if nfspvc.ObjectMeta.DeletionTimestamp != nil {
+		if controllerutil.ContainsFinalizer(&nfspvc, utils.NfsPvcDeletionFinalizer) {
+			err, pvcDeleted, pvDeleted := areResourceDeleted(ctx, nfspvc, k8sClient)
+			if err != nil {
+				return err, false
+			}
+			if pvDeleted && pvcDeleted {
+				return finalizer.Remove(ctx, nfspvc, k8sClient), true
+			} else if !pvDeleted && pvcDeleted {
+				// if only the pvc was deleted successfully, the reconcile function will be triggered again for the pv cleanup.
+				if err := cleanup(ctx, nfspvc, k8sClient); err != nil {
+					return err, false
+				}
+				pvName := nfspvc.Name + "-" + nfspvc.Namespace + "-pv"
+				return fmt.Errorf("the pv %s is not deleted yet, %w", pvName, FailedCleanupError), false
+			}
+			if err := cleanup(ctx, nfspvc, k8sClient); err != nil {
+				return err, false
+			}
+		} else {
+			return nil, true
+		}
+	}
+	return nil, false
+}
+
+// deleteResource gets resource to delete and delete that resource from the cluster.
+func deleteResource(ctx context.Context, resource client.Object, k8sClient client.Client) error {
+	if err := k8sClient.Delete(ctx, resource); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+	return nil
+}
+
+// cleanup deletes the pvc and the pv that related to the nfspvc.
+func cleanup(ctx context.Context, nfspvc danaiov1alpha1.NfsPvc, k8sClient client.Client) error {
```

Comment:
This function doesn't need to take `nfspvc` as a parameter, you can just pass the name and the namespace since it's all that's used from the object